skyportal/models.py

```diff
@@ -1645,39 +1652,170 @@ class Spectrum(Base):
+        fluxerr_colindex=2,
+    ):
         """Generate a `Spectrum` from an ascii file.
 
         Parameters
         ----------
-
         filename : str
-           The name of the ASCII file containing the spectrum.
-        obj_id : str
-           The name of the Spectrum's Obj.
-        instrument_id : int
-           ID of the Instrument with which this Spectrum was acquired.
-        observed_at : string or datetime
-           Median UTC ISO time stamp of the exposure or exposures in which the Spectrum was acquired."
+           The original name of the ASCII file containing the spectrum,
+           or the name of the file on disk to read. If `data` is not provided,
+           `from_ascii` tries to read this file and parse its contents. If `data`
+           is provided, the filename is used only for bookkeeping purposes; no
+           attempt is made to actually read the file.
+        data: str, optional
+           The content of the ASCII file. If this argument is
+           supplied, then no attempt is made to read the file `filename` from
+           disk. This permits this method to be used on raw, in-memory data
+           (e.g., the kind supplied in an HTTP request) in addition to
+           file-like objects, saving the need for any interaction with the
+           filesystem.
+        obj_id : str, optional
+           The id of the Obj that this Spectrum is of, if not present
+           in the ASCII header.
+        instrument_id : int, optional
+           ID of the Instrument with which this Spectrum was acquired,
+           if not present in the ASCII header.
+        observed_at : string or datetime, optional
+           Median UTC ISO time stamp of the exposure or exposures in which
+           the Spectrum was acquired, if not present in the ASCII header."
+        wave_colindex: integer, optional
+           The 0-based index of the ASCII column corresponding to the wavelength
+           values of the spectrum (default 0).
+        flux_colindex: integer, optional
+           The 0-based index of the ASCII column corresponding to the flux
+           values of the spectrum (default 1).
+        fluxerr_colindex: integer or None, optional
+           The 0-based index of the ASCII column corresponding to the flux error
+           values of the spectrum (default 2). If there are only 2 columns
+           in the input file this value will be ignored. If there are more than
+           2 columns in the input file, but none of them correspond to flux error
+           values, set this parameter to None.
 
         Returns
         -------
-
         spec : `skyportal.models.Spectrum`
            The Spectrum generated from the ASCII file.
 
         """
-        data = np.loadtxt(filename)
-        if data.shape[1] != 2:  # TODO support other formats
-            raise ValueError(f"Expected 2 columns, got {data.shape[1]}")
+
+        filename = Path(filename)
+
+        if data is None:
```

Comment:
If we do the`file` argument which accepts filename or file-like, then these two duplicate clauses will collapse.

Reply:
Thanks, this is much nicer.